core/test/TreasureGrabTests.java

```diff
@@ -0,0 +1,192 @@
+    ) {
+
+        Arrays.stream(moves).forEach(logic::movePlayer);
+        logic.applyShadowToField();
+
+        Assertions.assertEquals(
+                "\n" + shadowExpectation.trim(),
+                "\n" + encodeShadow(logic).trim()
+        );
+    }
+
+    private void tryMoves(
+            final Logic logic,
+            final String movesCode,
+            final String shadowExpectation
+    ) {
+        final Logic.MoveDirection[] moves = parseMoveSequence(movesCode);
+
+        tryMoves(logic, moves, shadowExpectation);
+    }
+
+    private Logic tryMoves(
+            final int fieldWidth,
+            final int fieldHeight,
+            final Logic.Pos playerPos,
+            final Logic.Pos treasurePos,
+            final String movesCode,
+            final String shadowExpectation
+    ) {
+        final Logic.CellType[][] field = genEmptyField(fieldWidth, fieldHeight);
+        field[playerPos.y][playerPos.x] = Logic.CellType.ENTRANCE;
+        field[treasurePos.y][treasurePos.x] = Logic.CellType.TREASURE;
+        final Logic logic = new Logic(field, Map.of());
+
+        makeAssertion(logic, fieldWidth, fieldHeight);
+
+        tryMoves(logic, movesCode, shadowExpectation);
+        return logic;
+    }
+
+    private void makeAssertion(
+            Logic logic,
+            final int fieldWidth,
+            final int fieldHeight
+    ) {
+        Assertions.assertNotEquals(fieldHeight, 0);
+        Assertions.assertNotEquals(fieldWidth, 0);
+
+        Assertions.assertEquals(1, logic.allThings().filter(x -> x.getValue() == Logic.ThingType.PLAYER).count());
+        Logic.Pos act = logic.allThings().filter(x -> x.getValue() == Logic.ThingType.PLAYER).findFirst().get().getKey();
+        Logic.Pos exp = searchPos(logic, Logic.CellType.ENTRANCE);
+        Assertions.assertEquals(exp, act);
+    }
+
+    @Test
+    public void testCoverSimple() {
+        Logic logic = tryMoves(
+                8, 8,
+                new Logic.Pos(1, 1),
+                new Logic.Pos(3, 3),
+                "RRDD",
+                """
+                        oooooooo
+                        ooxxoooo
+                        oooxoooo
+                        ooo$oooo
+                        oooooooo
+                        oooooooo
+                        oooooooo
+                        oooooooo
+                        """
+        );
+        boolean[][] shadow = shadowCoverage(logic, 8, 8);
+
+        tryMoves(
+                logic,
+                "DDUU",
+                """
+                        oooooooo
+                        ooxxoooo
+                        oooxoooo
+                        ooo$oooo
+                        oooooooo
+                        oooooooo
+                        oooooooo
+                        oooooooo
+                        """
+        );
+        boolean[][] newShadow = shadowCoverage(logic, 8, 8);
+
+        Assertions.assertEquals(shadow, newShadow);
```

Comment:
Это не делает то, что хотелось бы (не сравнивает массивы поэлементно). Чтобы добиться нужного эффекта, надо делать вот так:

```suggestion
       Assertions.assertTrue(Arrays.deepEquals(shadow, newShadow));
```

Это пока временная "затычка". IDEA 100% не сможет нормально переварить такой Assert. Надо будет в будущем что-нить получше придумать.